docs/pages/api.rst

```diff
@@ -145,7 +145,32 @@ Another example
 Methods (operations)
 --------------------
 
-.. note:: Operations always returns a new collection object, with the exception of ``all``, ``count``, ``current``, ``key``.
+Background
+~~~~~~~~~~
+
+Operations are pure functions which can be used to manipulate an iterator, either directly
+or through the ``Collection`` object.
+
+.. literalinclude:: code/operations/background.php
+  :language: php
+
+When used separately, operations typically return a PHP `Generator`_ or an `Iterator`_.
+When used as a ``Collection`` method, operations fall into three main categories based on the return type:
+
+1. Operations that return a ``scalar`` value. Currently, this includes:
+   ``Contains``, ``Every``, ``Falsy``, ``Has``, ``Key``, ``Match`` (or ``MatchOne``), ``Nullsy``, ``Truthy``.
```

Comment:
I think the explanation is overall really great and I wouldn't change a thing, except for the classification points here.

What I propose is to list the operation types as such:

1. Operations which returns a boolean: countains, every, falsy, has... etc etc
2. Operations which returns a Collection of Collection: span, partition
3. Operation which returns an array: all
4. Operation which returns a Collection: all the rest.

(_be carefull _Collection::key()_ might return anything, not only a scalar, but anything_)

Also, what do you think about:

* Updating the signatures of all the operations in the API page and add their return type ?
or
* Updating all operations and add a note about the return type ?

(_about that page, it's very sad we cannot generate the documentation automatically from the classes themselves!_)

Let me know what you think.

Reply:
@drupol thank you for the feedback, I've done some changes to the numbered list to incorporate your suggestions, with a couple tweaks. Let me know what you think of this new classification :).

Regarding updating the signatures of the operations in the docs -> it's something that I thought about today as well 😄 . I'll do it in a future PR for all of them

Reply:
Thanks all good now :)

Reply:
@drupol great, this is all ready on my side 😁